File: core/dotnet/src/serialization/JsonSerializationWriter.cs
using System;
using System.Linq;
using System.IO;
using System.Text.Json;
using Kiota.Abstractions.Serialization;
using System.Threading.Tasks;
using System.Collections.Generic;
using System.Reflection;
using KiotaCore.Extensions;

namespace KiotaCore.Serialization {
    public class JsonSerializationWriter : ISerializationWriter, IDisposable {
        private readonly MemoryStream stream = new MemoryStream();
        public readonly Utf8JsonWriter writer;
        public JsonSerializationWriter()
        {
            writer = new Utf8JsonWriter(stream);
        }
        public Stream GetSerializedContent() {
            writer.Flush();
            stream.Position = 0;
            return stream;
        }
        public void WriteStringValue(string key, string value) {
            if(value != null) { // we want to keep empty string because they are meaningfull
                if(!string.IsNullOrEmpty(key)) writer.WritePropertyName(key);
                writer.WriteStringValue(value);
            }
        }
        public void WriteBoolValue(string key, bool? value) {
            if(!string.IsNullOrEmpty(key) && value.HasValue) writer.WritePropertyName(key);
            if(value.HasValue) writer.WriteBooleanValue(value.Value);
        }
        public void WriteIntValue(string key, int? value) {
            if(!string.IsNullOrEmpty(key) && value.HasValue) writer.WritePropertyName(key);
            if(value.HasValue) writer.WriteNumberValue(value.Value);
        }
        public void WriteFloatValue(string key, float? value) {
            if(!string.IsNullOrEmpty(key) && value.HasValue) writer.WritePropertyName(key);
            if(value.HasValue) writer.WriteNumberValue(value.Value);
        }
        public void WriteDoubleValue(string key, double? value) {
            if(!string.IsNullOrEmpty(key) && value.HasValue) writer.WritePropertyName(key);
            if(value.HasValue) writer.WriteNumberValue(value.Value);
        }
        public void WriteGuidValue(string key, Guid? value) {
            if(!string.IsNullOrEmpty(key) && value.HasValue) writer.WritePropertyName(key);
            if(value.HasValue) writer.WriteStringValue(value.Value);
        }
        public void WriteDateTimeOffsetValue(string key, DateTimeOffset? value) {
            if(!string.IsNullOrEmpty(key) && value.HasValue) writer.WritePropertyName(key);
            if(value.HasValue) writer.WriteStringValue(value.Value);
        }
        public void WriteEnumValue<T>(string key, T? value) where T : struct, Enum {
            if(!string.IsNullOrEmpty(key) && value.HasValue) writer.WritePropertyName(key);
            if(value.HasValue) {
                if(typeof(T).GetCustomAttributes<FlagsAttribute>().Any())
                    writer.WriteStringValue(Enum.GetValues<T>()
                                            .Where(x => value.Value.HasFlag(x))
                                            .Select(x => Enum.GetName<T>(x))
                                            .Select(x => x.ToFirstCharacterLowerCase())
                                            .Aggregate((x, y) => $"{x},{y}"));
                else writer.WriteStringValue(value.Value.ToString().ToFirstCharacterLowerCase());
            }
        }
        public void WriteCollectionOfPrimitiveValues<T>(string key, IEnumerable<T> values) {
            if(values != null) { //empty array is meaningful
                if(!string.IsNullOrEmpty(key)) writer.WritePropertyName(key);
                writer.WriteStartArray();
                foreach(var collectionValue in values) {
                    switch(collectionValue) {
                        case bool v:
                            writer.WriteBooleanValue(v);
                        break;
                        case string v:
                            writer.WriteStringValue(v);
                        break;
                        case Guid v:
                            writer.WriteStringValue(v);
                        break;
                        case DateTimeOffset v:
                            writer.WriteStringValue(v);
                        break;
                        case int v:
                            writer.WriteNumberValue(v);
                        break;
                        case float v:
                            writer.WriteNumberValue(v);
                        break;
                        case double v:
                            writer.WriteNumberValue(v);
                        break;
                        default:
                            throw new InvalidOperationException($"unknown type for serialization {collectionValue.GetType().FullName}");
                    }
                }
                writer.WriteEndArray();
            }
        }
        public void WriteCollectionOfObjectValues<T>(string key, IEnumerable<T> values) where T : class, IParsable<T>, new() {
            if(values != null) { //empty array is meaningful
                if(!string.IsNullOrEmpty(key)) writer.WritePropertyName(key);
                writer.WriteStartArray();
                foreach(var item in values.Where(x => x != null)) {
                    writer.WriteStartObject();
                    item.Serialize(this);
                    writer.WriteEndObject();
                }
                writer.WriteEndArray();
            }
        }
        public void WriteObjectValue<T>(string key, T value) where T : class, IParsable<T>, new() {
            if(value != null) {
                if(!string.IsNullOrEmpty(key)) writer.WritePropertyName(key);
                writer.WriteStartObject();
                value.Serialize(this);
                writer.WriteEndObject();
            }
        }
        public void Dispose()
        {
            writer.Dispose();
        }
    }
}

File: core/typescript/src/serialization/jsonSerializationWriter.ts
import { Parsable, SerializationWriter } from "@microsoft/kiota-abstractions";
import { TextEncoder } from "util";
import { ReadableStream } from 'web-streams-polyfill/es2018';

export class JsonSerializationWriter implements SerializationWriter {
    private readonly writer: string[] = [];
    private static propertySeparator = `,`;
    public writeStringValue = (key?: string, value?: string): void => {
        key && value && this.writePropertyName(key);
        value && this.writer.push(`"${value}"`);
        key && value && this.writer.push(JsonSerializationWriter.propertySeparator);
    }
    private writePropertyName = (key: string) : void => {
        this.writer.push(`"${key}":`);
    }
    public writeBooleanValue = (key?: string, value?: boolean): void => {
        key && value && this.writePropertyName(key);
        value && this.writer.push(`${value}`);
        key && value && this.writer.push(JsonSerializationWriter.propertySeparator);
    }
    public writeNumberValue = (key?: string, value?: number): void => {
        key && value && this.writePropertyName(key);
        value && this.writer.push(`${value}`);
        key && value && this.writer.push(JsonSerializationWriter.propertySeparator);
    }
    public writeGuidValue = (key?: string, value?: string): void => {
        key && value && this.writePropertyName(key);
        value && this.writer.push(`"${value}"`);
        key && value && this.writer.push(JsonSerializationWriter.propertySeparator);
    }
    public writeDateValue = (key?: string, value?: Date): void => {
        key && value && this.writePropertyName(key);
        value && this.writer.push(`"${value.toISOString()}"`);
        key && value && this.writer.push(JsonSerializationWriter.propertySeparator);
    }
    public writeCollectionOfPrimitiveValues = <T>(key?: string, values?: T[]): void => {
        if(values) {
            key && this.writePropertyName(key);
            this.writer.push(`[`);
            values.forEach(v => {
                if(v instanceof Boolean) {
                    this.writeBooleanValue(undefined, v as any as boolean);
                } else if (v instanceof String) {
                    this.writeStringValue(undefined, v as any as string);
                } else if (v instanceof Date) {
                    this.writeDateValue(undefined, v as any as Date);
                } else if (v instanceof Number) {
                    this.writeNumberValue(undefined, v as any as number);
                } else {
                    throw new Error(`encountered unknown value type during serialization ${typeof v}`);
                }
            });
            this.writer.push(`]`);
        }
    }
    public writeCollectionOfObjectValues = <T extends Parsable<T>>(key?: string, values?: T[]): void => {
        if(values) {
            key && this.writePropertyName(key);
            this.writer.push(`[`);
            values.forEach(v => {
                this.writeObjectValue(undefined, v);
                this.writer.push(JsonSerializationWriter.propertySeparator);
            });
            if(values.length > 0) { //removing the last separator
                this.writer.pop();
            }
            this.writer.push(`]`);
        }
        key && values && this.writer.push(JsonSerializationWriter.propertySeparator);
    }
    public writeObjectValue = <T extends Parsable<T>>(key?: string, value?: T): void => {
        if(value) {
            if(key) {
                this.writePropertyName(key);
            }
            this.writer.push(`{`);
            value.serialize(this);
            if(this.writer.length > 0 && this.writer[this.writer.length - 1] === JsonSerializationWriter.propertySeparator) { //removing the last separator
                this.writer.pop();
            }
            this.writer.push(`}`);
        }
    }
    public writeEnumValue = <T>(key?: string | undefined, ...values: (T | undefined)[]): void => {
        if(values.length > 0) {
            const rawValues = values.filter(x => x !== undefined).map(x => `${x}`);
            if(rawValues.length > 0) {
                this.writeStringValue(key, rawValues.reduce((x, y) => `${x}, ${y}`));
            }
        }
    }
    public getSerializedContent = (): ReadableStream<any> => {
        const encoded = new TextEncoder().encode(this.writer.join(""));
        return new ReadableStream<Uint8Array>({
            start: (controller) => {
                controller.enqueue(encoded);
                controller.close();
            }
        });
    }
}